File: poa-search-api/src/main/java/nl/rabobank/service/poa/search/detailedpowerofattorney/DetailedPowerOfAttorney.java
package nl.rabobank.service.poa.search.detailedpowerofattorney;


import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import io.swagger.annotations.ApiModelProperty;
import org.immutables.value.Value;

import javax.annotation.Nonnull;
import java.util.List;


@Value.Immutable
@JsonDeserialize(as = ImmutableDetailedPowerOfAttorney.class)
public interface DetailedPowerOfAttorney {

    @Nonnull
    @ApiModelProperty(required = true, value = "Id of Power of Attorney.")
    String getId();

    @Nonnull
    @ApiModelProperty(required = true, value = "Grantor name of Power of Attorney")
    String getGrantor();

    @Nonnull
    @ApiModelProperty(required = true, value = "Grantee name of Power of Attorney")
    String getGrantee();

    @Nonnull
    @ApiModelProperty(required = true, value = "Information of the account that has been granted by Power of Attorney")
    Account getAccount();

    @Nonnull
    @ApiModelProperty(required = true, value = "Direction of Power of Attorney", allowableValues = "GIVEN, RECEIVED")
    Direction getDirection();

    @Nonnull
    @ApiModelProperty(required = true, value = "List of authorizations for Power of Attorney")
    List<Authorization> getAuthorizations();

    @Nonnull
    @ApiModelProperty(required = true, value = "List of cards that have been granted by Power of Attorney")
    List<Card> getCards();

}

File: poa-search-server/src/main/kotlin/nl/rabobank/service/poa/search/detailedpowerofattorney/model/DetailedPowerOfAttorney.kt
package nl.rabobank.service.poa.search.detailedpowerofattorney.model

import nl.rabobank.service.poa.search.account.model.Account
import nl.rabobank.service.poa.search.card.creditcard.model.CreditCard
import nl.rabobank.service.poa.search.card.debitcard.model.DebitCard
import nl.rabobank.service.poa.search.detailedpowerofattorney.ImmutableDetailedPowerOfAttorney
import nl.rabobank.service.poa.search.powerofattorney.model.Authorization
import nl.rabobank.service.poa.search.powerofattorney.model.Direction
import nl.rabobank.service.poa.search.powerofattorney.model.expose
import nl.rabobank.service.poa.search.util.Exposable
import nl.rabobank.service.poa.search.detailedpowerofattorney.DetailedPowerOfAttorney as ExposedClientPoaDetail

data class DetailedPowerOfAttorney(
        val id: String,
        val grantor: String,
        val grantee: String,
        val direction: Direction,
        val authorizations: List<Authorization>,
        val account: Account,
        val creditCards: List<CreditCard>,
        val debitCards: List<DebitCard>
): Exposable<ExposedClientPoaDetail> {
    override fun expose()= ImmutableDetailedPowerOfAttorney.builder()
            .id(id)
            .grantor(grantor)
            .grantee(grantee)
            .account(account.expose())
            .direction(direction.expose())
            .addAllAuthorizations(authorizations.map{it.expose()})
            .addAllCards((creditCards.map { it.expose() } + debitCards.map { it.expose() }))
            .build()
}